### PR TITLE
Configure Firebase Storage Integration

### PR DESCRIPTION
**Pull request by Alvaro747**
- Set up storage configuration to utilize Firebase.
- Implemented file upload functionality.
- Created routes for handling file uploads.
- Established services for Firebase integration and file operations.

### CODE REVIEW

File: src/core/api.manager.js [hunk 1]
```diff
@@ -11,34 +11,77 @@ class ApiManager {
...
     this._swaggerJsdoc = dependencies.swaggerJsdoc;
     this._swaggerUi = dependencies.swaggerUi;
+    this._storage = {};
 
     /* Assigments */
     this._namespace = '[Server]::[API]::[Manager]';
     this._apiRoutes = this._express.Router();
     this._path = dependencies.path;
+    this._multer = dependencies.multerModule;
+    this._storage = {};
```
Comment (thEpisode) [inline src/core/api.manager.js:20]: is duplicated the variable _storage?

File: config/template.json [hunk 2]
```diff
@@ -66,6 +66,22 @@
...
+        "apiKey": "xxxx__your_config",
+        "authDomain": "xxxx__your_config",
+        "projectId": "xxxx__your_config",
+        "storageBucket": "xxxx__your_config",
+        "messagingSenderId": "xxxx__your_config",
+        "appId": "xxxx__your_config"
+      },
+      "SETTINGS": {
+        "timestampsInSnapshots": true,
+        "fileSize": 5242880
```
Comment (thEpisode) [inline config/template.json]: change to maxFileSize

File: src/core/storage.manager.js [hunk 3]
```diff
@@ -1,88 +1,89 @@
...
 
-  async firebaseConfig() {
+  #setupSelectedStorageSource() {
     try {
-      this._stg = this._dependencies.firebase.storage();
+      const DataSource = require(
+        `${this._dependencies.root}/src/storage-source/${this._currentStorageSourceConfig.path}`,
+      );
+
+      this._stg.firebase =
```
Comment (thEpisode) [inline src/core/storage.manager.js]: you cannot assign any firebase dependency in core

File: src/services/upload/upload.service.js [hunk 4]
```diff
@@ -19,38 +19,31 @@ class UploadService {
   async uploadFile(req) {
     try {
       if (!req || !req.file) {
-        return this._utilities.io.response.error('Add a file');
+        return this._utilities.io.response.error('Agrega un archivo');
```
Comment (thEpisode) [inline src/services/upload/upload.service.js]: English strings

File: src/services/upload/upload.service.js [hunk 5]
```diff
@@ -19,38 +19,31 @@ class UploadService {
...
-
-      if (
-        !this._services[req.body.route] ||
-        !this._services[req.body.route][req.body.handler]
-      ) {
-        return this._utilities.io.response.error(
-          'Given path to handle your bulk request is not available',
-        );
+      if (!req || !req.body || !req.body.folder) {
+        return this._utilities.io.response.error('Provea un folder, por favor');
```
Comment (thEpisode) [inline src/services/upload/upload.service.js]: English

File: src/storage-source/base/data-source.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
+class DataSource {
+  constructor(dependencies) {
+    /* Base Properties */
+    this._dependencies = dependencies;
+    this._utilities = this._dependencies.utilities;
+    this._console = this._dependencies.console;
+  }
+
+  async create({ tableName, entity } = {}) {
```
Comment (thEpisode) [inline src/storage-source/base/data-source.js]: why you need a create?

File: src/storage-source/base/data-source.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    if (!entity) {
+      return this._utilities.io.response.error('Provide an entity');
+    }
+
+    return this._utilities.io.response.success({
+      id: this._utilities.generator.id({ length: 15 }),
+    });
+  }
+
+  async update({ tableName, entity } = {}) {
```
Comment (thEpisode) [inline src/storage-source/base/data-source.js]: why you need an update?

File: src/storage-source/base/data-source.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    if (!entity) {
+      return this._utilities.io.response.error('Provide an entity');
+    }
+
+    return this._utilities.io.response.success('Valid table name');
+  }
+
+  async getByFilters({ tableName } = {}) {
```
Comment (thEpisode) [inline src/storage-source/base/data-source.js]: why you need a getByFilters?

File: src/storage-source/firebase/firebase-storage-source.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    this._AdminFirestore =
+      this._dependencies.config.DATASOURCE_CONFIGS.FIRESTORE.CONNECTION_OBJ;
+    this._StorageConnectionObj = this._StoreSourceConfig.CONNECTION_OBJ || {};
+    this._databaseSettings = this._StoreSourceConfig.SETTINGS || {};
+  }
+
+  async setup() {
+    // Setup the driver/client
+    // configurate and initialize firebase admin
+
```
Comment (thEpisode) [inline src/storage-source/firebase/firebase-storage-source.js:24]: Remove extra spaces

File: src/storage-source/firebase/firebase-storage-source.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    try {
+      const bucketName = this._StorageConnectionObj.storageBucket;
+      const bucket = this._storage.bucket(bucketName);
+      const bucketFile = bucket.file(`${folder}/${clientFile.originalname}`);
+      const uploadParams = this.#getUploadParams(clientFile);
+
+      await bucketFile.save(clientFile.buffer, uploadParams);
+      const url = await bucketFile.getSignedUrl({
+        action: 'read',
+        expires: '01-01-2050',
```
Comment (thEpisode) [inline src/storage-source/firebase/firebase-storage-source.js]: Assign as a variable, first get as an argument and default value is the 2050

**Review by camilo-appsky — APPROVED** [review]

**Review by camilo-appsky — APPROVED** [review]

**Review by thEpisode — APPROVED** [review]

**Pull request merged**